Data access api README/Docs update

anamanica opened the pull request:
# Change Summary

## Overview
<!--Add a list or paragraph giving an overview of your changes-->
This PR combines the information in the README for the data access api into the docs.

(I will have to cut down the README in a PR on the `imap-data-access` repo.)

## New Dependencies
<!--List any new dependencies introduced by these changes-->
None

## New Files
<!--List each new file and add bullets to describe the purpose/function of the new file.
This should be high level, but enough detail for the reviewer to understand what purpose(s) the
code in the file is serving-->
None

## Deleted Files
<!--List each deleted file and add one or more bullets with the rationale for why the file is being deleted-->
None

## Updated Files
<!--List each updated file and add bullets to describe the purpose/function of the updates.
This should be high level, but enough detail for the reviewer to understand what purpose(s) the
updated code in the file is serving-->
- index.rst
   - adding in README information

## Testing
None

anamanica committed:
Doc changes

anamanica committed:
Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api

Updating branch

anamanica committed:
Updating

anamanica committed:
Fixing

anamanica committed:
Fixing look

anamanica committed:
Playing with format

anamanica committed:
Fixing again

anamanica committed:
Fixing again

anamanica committed:
Fixing again

anamanica committed:
Fixing again

anamanica committed:
Fixing again

anamanica committed:
Fixing again

anamanica committed:
Updating Branch
Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api

anamanica committed:
Small change

anamanica committed:
More formatting

anamanica committed:
Updating branch

Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api

anamanica committed:
PR updates

anamanica committed:
I changed my mind on format

anamanica committed:
Inital Changes

anamanica committed:
Some small changes

anamanica committed:
Little changes

anamanica committed:
Fixing Merge

anamanica committed:
Fixing docs error

bourque commented:
Adding @greglucas as a reviewer here since he wrote most of this documentation and may have opinions about this!

bourque commented:
Pertains to https://github.com/IMAP-Science-Operations-Center/imap-data-access/issues/75

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,44 @@ that the full implementation of the functionality is yet to be completed.*
 
 The API can be accessed from the following URL [WIP]: https://api.dev.imap-mission.com
 
+To Install
+==========
```
```suggestion
To Install
----------
```

I think these section headers need to be changed to 'subsections'. I notice that the way these are currently defined is messing up the Table of Contents on the main landing page.

greglucas reviewed (approved):
I don't have any major opinions on this, so looks good to me.

The one concern I do have is that it'd be nice if we could have one "source" instead of copying the content to two places because they are bound to get out of sync at some point. Is there a way to `.. include:: https://path/to/external_repo/README.md` or something similar to that for referencing an external project file?

bourque reviewed (commented):
Thank you for moving this documentation over! Overall this is looking good, but now that I see it all in one place, I am finding the layout a bit confusing and too intermingled with the existing OpenAPI specification part of the documentation.  As such, I have a suggestion for a better layout:

- Data Access API (i.e. the introduction as you currently have it)
  - Command Line Utility
    - To Install 
    - Base Command Argumments
    - Query/search for data (with example from README)
    - Download a file (with example from README)
    - Upload a file (with example from README)
  - Importing as a package (with example from README) 
  - Configuration
    - Data Directory
    - Data Access URL
  - File Validation  
  - Troubleshooting
  - REST API Specification
    - Query OpenAPI specs
    - Download OpenAPI specs
    - Upload API specs

bourque commented:
> I don't have any major opinions on this, so looks good to me.
> 
> The one concern I do have is that it'd be nice if we could have one "source" instead of copying the content to two places because they are bound to get out of sync at some point. Is there a way to `.. include:: https://path/to/external_repo/README.md` or something similar to that for referencing an external project file?

@greglucas Totally agree, and I think the end goal is to have all of the documentation pertaining to the Data Access API in this RtD page, and then basically have the `imap-data-access` repo `README` just say something like "For documentation how to use this, go to the official docs here". Does that sound good to you?

anamanica committed:
Updating Branch

Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api-README

greglucas commented:
> @greglucas Totally agree, and I think the end goal is to have all of the documentation pertaining to the Data Access API in this RtD page, and then basically have the imap-data-access repo README just say something like "For documentation how to use this, go to the official docs here". Does that sound good to you?

I would personally prefer the opposite direction 🤷 I think a package should be standalone if possible and the README should describe whats going on within that package without needing to click somewhere else. I just don't know how feasible any of it is to go either direction.

anamanica committed:
Updating Branch

Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api-README

anamanica committed:
Updating Branch.

Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api-README

anamanica committed:
Order changes

anamanica committed:
Order changes

anamanica committed:
Pre-commit

anamanica committed:
Small changes

anamanica committed:
Updating Branch.

Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api-README

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,195 @@ that the full implementation of the functionality is yet to be completed.*
 
 The API can be accessed from the following URL [WIP]: https://api.dev.imap-mission.com
 
+Command Line Utility
+====================
+To Install
+----------
```
```suggestion
Command Line Utility
--------------------
To Install
^^^^^^^^^^
```

It looks like these headings (and throughout this document) are still messing up the main RtD page layout. I think these need to be changed to another level 'down' in markdown.

anamanica replied:
Oh! Sorry, Okay. I think I understand what you are saying now! Should I do this with every heading?

bourque replied:
Yes, except for the very first one (line 4), which is fine as is.

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,195 @@ that the full implementation of the functionality is yet to be completed.*
+
+Add the -h flag with any base command for more information on use and functionality.
+
+.. openapi:: openapi.yml
+   :group:
+   :include: /query
```
This part is describing the REST API specification, so it should be moved to under that section.

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,195 @@ that the full implementation of the functionality is yet to be completed.*
+    [{'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}, {'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': ' ...
+
+
+.. openapi:: openapi.yml
+   :group:
+   :include: /download
```
Similar to the last comment, this is describing the REST API so should also be moved (same with the `/upload` section.

anamanica committed:
More PR comments

anamanica committed:
Adding descriptions

anamanica committed:
Hopefully fixing my mistake

anamanica committed:
More small changes

bourque commented on docs/source/data-access-api/index.rst:
Since this document is quite lengthy now and have several sections and subsections, I suggest to add a table of contents somewhere towards the top. (Could be done in a future PR if desired)

bourque commented on docs/source/data-access-api/index.rst:
Since this document is quite lengthy now and have several sections and subsections, I suggest to add a table of contents somewhere towards the top. (Could be done in a future PR if desired)

bourque replied:
Nevermind, Ana pointed out to me that there is a built-in table of contents on ReadTheDocs, which I totally missed!

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -4,10 +4,10 @@ Data Access API
 
 The `imap-data-access <https://github.com/IMAP-Science-Operations-Center/imap-data-access>`_
-repository provides programmatic access and a command-line utility for
+repository provides programmatic access, and a command-line utility for
 interacting with the API. It is the preferred way to use the API.
```
```suggestion
repository provides a command-line utility and python package for
interacting with the API programmatically. It is the preferred way to use the API.
```

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -4,10 +4,10 @@ Data Access API
 interacting with the API. It is the preferred way to use the API.
 
-The SDC provides a REST API that allows users to upload and download files, as
+The SDC also provides a REST API that allows users to upload and download files as
```
I suggest to change this whole paragraph to something like this:

```suggestion
Users may also download, upload, and query via the REST API directly through the browser or `curl` commands. The [REST API Specification](link) section below describes the various endpoints that are supported and how to use them.
```

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+
+        imap-data-access query --start-date 20240101 --end-date 20241231 --output-format json
+        # The following line is returned:
+        [{'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension': 'pkts'}, {'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extension ...
```
Could you split this up into multiple lines? The way this is displayed on the page (one long line that you have to scroll through) is a bit difficult to see.

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+Download
+^^^^^^^^
+
+To download files using the CLI tool, use the command ``download``. These will be placed in a ``data`` directory.
```
```suggestion
To download files using the CLI tool, use the command ``download``. The downloaded files will be placed in a ``data`` directory.
```

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+accordingly to suit your desires.
+
+When downloading a file from the API, different folders within the ``data`` directory will be made to better
+organize the downloaded files. See the example path: ``data/imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_20240105_v00-01.pkts``
```
```suggestion
organize the downloaded files. See the example path: ``data/imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_20240105_v00-01.pkts``. The ``data`` directory and its structure is futher described [here](<link to data directory section>)
```

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+        # Search for files
+        results = imap_data_access.query(instrument="mag", data_level="l0")
+        # results is a list of dictionaries
+        # [{'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105','version': 'v001', 'extension': 'pkts'}, {'file_path': 'imap/swe/l0/2024/01/imap_swe_l0_sci_20240105_v001.pkts', 'instrument': 'swe', 'data_level': 'l0', 'descriptor': 'sci', 'start_date': '20240105', 'version': 'v001', 'extensio ...
```
Similar to my comment above, if this could be split into multiple lines I think that would read better

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+^^^^^^^^^^^^^^
+
+The folder structure for data files within the IMAP SDC is rigidly defined, so the data access will mimic that structure to make sure all data is stored in the same hierarchical structure as the SDC. This will enable seamless transition between a user's local system and the SDC. This is only used for downloads.
+A user's root data location can be specified as an environment variable IMAP_DATA_DIR or through a configuration dictionary within the package itself imap_data_access.config["DATA_DIR"]. If the IMAP_DATA_DIR variable is not set, the program defaults to the user's current working directory + data/.
```
```suggestion
A user's root data location can be specified as an environment variable ``IMAP_DATA_DIR`` or through a configuration dictionary within the package itself (``imap_data_access.config["DATA_DIR"]``). If the ``IMAP_DATA_DIR`` variable is not set, the program defaults to the user's current working directory + ``data/``.
```

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+Data Access URL
+^^^^^^^^^^^^^^^
+
+To change the default URL that the package accesses, you can set the environment variable ``IMAP_DATA_ACCESS_URL`` or within the package ``imap_data_access.config["DATA_ACCESS_URL"]``. The default is the development server https://api.dev.imap-mission.com.
```
```suggestion
To change the default URL that the package accesses, you can set the environment variable ``IMAP_DATA_ACCESS_URL`` or within the package ``imap_data_access.config["DATA_ACCESS_URL"]``. The default is the development server (``https://api.dev.imap-mission.com``).
```

bourque commented on docs/source/data-access-api/index.rst:
```diff
@@ -16,21 +16,201 @@ that the full implementation of the functionality is yet to be completed.*
+File Validation
+---------------
+
+This package validates filenames and paths to check they follow our standards, as defined by the filename conventions. There is also a class available for use by other packages to create filepaths and filenames that follow the IMAP SDC conventions.
```
```suggestion
This package validates filenames and paths to check they follow our standards, as defined by the [filename conventions](https://imap-processing.readthedocs.io/en/latest/development-guide/style-guide/naming-conventions.html). There is also a class available for use by other packages to create filepaths and filenames that follow the IMAP SDC conventions.
```

bourque reviewed (changes requested):
Thanks for these updates, this is looking really good! As Greg suggested, I think it is worth exploring if we can automatically include the contents of the `imap-data-access` `README` file in here, but that can be explored in a separate PR if you'd like.

I just have a few suggestions to help clarify some areas or make them a bit more readable.

anamanica committed:
Updating branch.

Merge branch 'dev' of github.com:IMAP-Science-Operations-Center/imap_processing into data-access-api-README

anamanica committed:
Testing ref links

anamanica committed:
Testing ref links AGAIN

anamanica committed:
More ref testing

anamanica committed:
Checking formatting

anamanica committed:
Small changes

anamanica committed:
Small change

anamanica committed:
Final comment changes

anamanica committed:
Final touches

bourque reviewed (approved):
Thanks for making those updates, this looks good to go!

bourque commented:
@anamanica Just a heads up that you will likely run into merge conflicts once #844 is merged, since that PR is changing the directory names a bit.

maxinelasp commented:
@anamanica are you planning on merging this soon?

greglucas commented:
@anamanica, are you going to come back to this, or did you want someone else to try and pick it up?

greglucas commented:
Closing this for lack of response. I think we can re-open it later if need be or another branch can be opened because there are conflicts.
